### PR TITLE
feat: release script accepts new_version arg

### PR DESCRIPTION
**Pull request by devjgm**
This PR teaches the release.sh script how to deal with version numbers
that don't start with a leading 0. This is important since Spanner is
now 1.0.0. release.sh now accepts an optional "new_version" argument to
let the user specify the desired new version number on the command line
rather than computing it from previous tags. This is useful when we want
to jump to a non-sequential version number. For example, this would have
worked for the Spanner 1.0.0 release we did a few days ago.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp-common/264)

### CODE REVIEW

File: release/release.sh [hunk 1]
```diff
@@ -93,19 +117,38 @@ if ! command -v hub > /dev/null; then
...
+  NEW_VERSION="${VERSION_ARG}"
+  echo "New version: ${NEW_VERSION}"
+else
+  # Figures out the most recent tagged version, and computes the next version.
+  readonly TAG="$(git describe --tags --abbrev=0 origin/master)"
+  readonly CUR_TAG="$(test -n "${TAG}" && echo "${TAG}" || echo "v0.0.0")"
+  readonly CUR_VERSION="${CUR_TAG#v}"
+  NEW_VERSION="$(perl -pe 's/(\d+).(\d+).(\d+)/"$1.${\($2+1)}.$3"/e' <<<"${CUR_VERSION}")"
+  echo "Current tag: ${CUR_TAG}"
+  echo "New version: ${NEW_VERSION}"
```
Comment (devbww) [inline release/release.sh]: Perhaps report tag+tag, or version+version, but not tag+version?

Comment (devjgm) [reply release/release.sh]: Good suggestion. I changed it so the banner now prints something like `"release info for ${CUR_TAG} -> ${NEW_TAG}"`, both tags

**Comment by devjgm** [comment]
Ping

File: release/release.sh [hunk 2]
```diff
@@ -1,21 +1,38 @@
...
+#   # NO CHANGES ARE PUSHED. Shows what commands would be run.
+#   $ release.sh googleapis/google-cloud-cpp-spanner 2.0.0
+#
+#   # PUSHES CHANGES to release -spanner
 #   $ release.sh -f googleapis/google-cloud-cpp-spanner
 #
+#   # PUSHES CHANGES to release -spanner, setting its new version to 2.0.0
+#   $ release.sh -f googleapis/google-cloud-cpp-spanner 2.0.0
+#
 # This script creates a "release" on github by doing the following:
```
Comment (devbww) [inline release/release.sh]: Perhaps this should come before the examples?

File: release/release.sh [hunk 3]
```diff
@@ -1,21 +1,38 @@
...
+#                                  new version will be computed from existing
+#                                  git tags. This flag should only be needed
+#                                  when jumping to a non-sequential version
+#                                  number.
 #
 #   Options:
 #     -f     Force; actually make and push the changes
 #     -h     Print help message
 #
 # Example:
```
Comment (devbww) [inline release/release.sh]: s/Example/Examples/

File: release/release.sh [hunk 4]
```diff
@@ -1,21 +1,38 @@
 #!/bin/bash
 #
 # Usage:
-#   $ release.sh [-f] <organization/project-name>
+#   $ release.sh [-f] <organization/project-name> [new-version]
```
Comment (devbww) [inline release/release.sh]: s/new-version/<new-version>/ perhaps?

File: release/release.sh [hunk 5]
```diff
@@ -47,15 +64,23 @@ done
...
+PROJECT_ARG=""
+VERSION_ARG=""
+if [[ $# -eq 1 ]]; then
+  PROJECT_ARG="$1"
+elif [[ $# -eq 2 ]]; then
+  PROJECT_ARG="$1"
+  VERSION_ARG="$2"
+else
+  echo "Invalid arguments"
   echo "$USAGE"
```
Comment (devbww) [inline release/release.sh]: 1>&2 perhaps?

File: release/release.sh [hunk 6]
```diff
@@ -93,19 +118,36 @@ if ! command -v hub > /dev/null; then
...
+  NEW_VERSION="${VERSION_ARG}"
+else
+  # No new version specified; compute the new version number
+  NEW_VERSION="$(perl -pe 's/(\d+).(\d+).(\d+)/"$1.${\($2+1)}.$3"/e' <<<"${CUR_VERSION}")"
+fi
+declare -r NEW_VERSION
+
+# Avoid handling patch releases for now, because we wouldn't need a new branch
+# for those.
+if ! grep -P "\d+\.\d+\.0" <<<"${NEW_VERSION}" > /dev/null; then
```
Comment (devbww) [inline release/release.sh]: Opt: Consider `case` builtin?
```
case "$1" in
[1-9][0-9]*.[1-9][0-9]*.0)
  :;;
*)
  <error exit>;;
esac
```

File: release/release.sh [hunk 7]
```diff
@@ -93,19 +118,36 @@ if ! command -v hub > /dev/null; then
...
+  echo "Sorry, cannot handle patch releases (yet)"
+  echo "$USAGE"
```
Comment (devbww) [inline release/release.sh]: 1>&2 perhaps?

File: release/release.sh (hunk 4 above)
Comment (devbww) [reply release/release.sh]: Oops ... forgot to escape angled-brackets.
Should have read ... `s/new-version/<new-version>/`

Comment (devjgm) [reply release/release.sh]: Done

File: release/release.sh (hunk 3 above)
Comment (devjgm) [reply release/release.sh]: Done.

File: release/release.sh (hunk 2 above)
Comment (devjgm) [reply release/release.sh]: Done.

File: release/release.sh (hunk 5 above)
Comment (devjgm) [reply release/release.sh]: Done, here and elsewhere.

File: release/release.sh (hunk 6 above)
Comment (devjgm) [reply release/release.sh]: I think the `if` with grep regex is clearer, IMO, and would prefer to keep it.

File: release/release.sh (hunk 7 above)
Comment (devjgm) [reply release/release.sh]: Done.

**Pull request merged**